frontend/model/contracts/chatroom.js

```diff
@@ -0,0 +1,372 @@
+      return {
+        createdDate: new Date().toISOString(),
+        username,
+        identityContractID
+      }
+    }
+  },
+  state (contractID) {
+    return sbp('state/vuex/state')[contractID]
+  },
+  getters: {
+    currentChatRoomState (state) {
+      return state
+    },
+    chatRoomSettings (state, getters) {
+      return getters.currentChatRoomState.settings || {}
+    },
+    chatRoomAttributes (state, getters) {
+      return getters.currentChatRoomState.attributes || {}
+    },
+    chatRoomUsers (state, getters) {
+      return getters.currentChatRoomState.users || {}
+    },
+    chatRoomLatestMessages (state, getters) {
+      const messages = getters.currentChatRoomState.messages || []
+      return getLatestMessages({
+        count: getters.chatRoomSettings.messagesPerPage,
+        messages
+      })
+    }
+  },
+  actions: {
+    // This is the constructor of Chat contract
+    'gi.contracts/chatroom': {
+      validate: chatRoomType,
+      process ({ meta, data }, { state }) {
+        const initialState = merge({
+          settings: {
+            messagesPerPage: CHATROOM_MESSAGES_PER_PAGE,
+            maxNameLetters: CHATROOM_NAME_LIMITS_IN_CHARS,
+            maxDescriptionLetters: CHATROOM_DESCRIPTION_LIMITS_IN_CHARS
+          },
+          users: {},
+          messages: []
+        }, {
+          attributes: {
+            ...data,
+            creator: meta.username,
+            deletedDate: null,
+            archievedDate: null
+          }
+        })
+        for (const key in initialState) {
+          Vue.set(state, key, initialState[key])
+        }
+      },
+      sideEffect ({ contractID, meta }) {
+        const rootState = sbp('state/vuex/state')
+        if (meta.username === rootState.loggedIn.username) {
+          sbp('state/vuex/commit', 'setCurrentChatRoomId', { chatRoomId: contractID })
+        }
+      }
+    },
+    'gi.contracts/chatroom/join': {
+      validate: objectMaybeOf({
+        username: string,
+        referer: string
+      }),
+      process ({ data, meta, hash }, { state }) {
+        const { username, referer } = data
+
+        const notificationType = username === referer ? MESSAGE_NOTIFICATIONS.JOIN_MEMBER : MESSAGE_NOTIFICATIONS.ADD_MEMBER
+        const notificationData = createNotificationData(
+          notificationType,
+          notificationType === MESSAGE_NOTIFICATIONS.ADD_MEMBER ? { username } : {}
+        )
+        const newMessage = createMessage({ meta, hash, data: notificationData, state })
+        Vue.set(state.messages, [state.messages.length], newMessage)
+
+        if (state.users[username] && !state.users[username].departedDate) {
+          console.log(`chatroom Join: ${username} is already joined the chatroom #${state.name}`)
+          return
+        }
+
+        Vue.set(state.users, username, {
+          joinedDate: meta.createdDate,
+          departedDate: null
+        })
```

Comment:
Can you explain this `departedDate` thing? Is it necessary for the UI?

Reply:
Once the user leaves the chatroom, `departedDate` is set.
Using this value, we decide if the user has joined or left.

Reply:
Why can't we decide that simply by their presence or absence from the room?

Reply:
Even though they are part of a group, they can be absent in any chatrooms inside the group.
Those `joinedDate` and `departedDate` fields are used to decide if a user is joined or left a specific chatroom.
Additionally, even though a user left the chatroom, his username is stored in `state.users`. In this case, his `departedDate` value is set to the datetime when he left the chatroom.

Reply:
>  Additionally, even though a user left the chatroom, his username is stored in state.users. In this case, his departedDate value is set to the datetime when he left the chatroom.

Yes, so my question is, why not simply remove them from `state.users` when they leave the room? Wouldn't that be simpler and result in less data needing to be stored? Why keep them around in `state.users`?

Reply:
Actually, I was going to use `departedDate` for **maybe** future functions like the following.
- Banned user list.
- Restrict to joining in a short period after leaving ( Block Spam )

I guess we don't definitely need `departedDate` for the `Banned List`. We can make `state.bannedUsers` field instead of `departedDate` inside `state.users`
And also I am not sure we will need those two features later.

Reply:
One thing I need to say is that we need to use the `group.profiles` to get the message sender details.
That's because `userA` can leave the group after joining and sending some messages.